src/Emulator/Peripherals/Peripherals/MTD/OpenTitan_FlashController.cs

```diff
@@ -0,0 +1,1185 @@
+                opStatusRegisterErrorFlag.Value |= true;
+            }
+        }
+
+        private void DoEraseInfoPage(long pageAddress){
+            /*
+            ** TODO(julianmb): memory protection should be implemented.
+            ** It currently doesnt have a unit test though.
+            ** BANK*_INFO*_PAGE_CFG_* registers are of interest
+            */
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoEraseInfoPage: pageAddress = 0x{0:X}",
+                pageAddress);
+            long pageOffset = pageAddress - FLASH_MEM_BASE_ADDR;
+            long bytesPerPage = FLASH_WORDS_PER_PAGE * FLASH_WORD_SZ;
+            for(long i = 0 ; i < bytesPerPage ; i++){
+                infoFlash.WriteByte(pageOffset + i, 0xff);
+            }
+        }
+
+        private void DoReadData(long busWordAddress){
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = flash.ReadDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: value = 0x{0:X}",
+                value);
+            if(IsOperationAllowed(OperationType.READ_DATA, busWordAddress)){
+                readFifo.Enqueue(value);
+            }else{
+                interruptStatusRegisterOpErrorFlag.Value |= true;
+                opStatusRegisterErrorFlag.Value |= true;
+            }
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadData: readFifo.Count = {0}",
+                readFifo.Count);
+        }
+
+        private void DoReadInfo(long busWordAddress){
+            /*
+            ** TODO(julianmb): memory protection should be implemented.
+            ** It currently doesnt have a unit test though.
+            ** BANK*_INFO*_PAGE_CFG_* registers are of interest
+            */
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = infoFlash.ReadDoubleWord(busWordAddress - FLASH_MEM_BASE_ADDR);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: value = 0x{0:X}",
+                value);
+            readFifo.Enqueue(value);
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoReadInfo: readFifo.Count = {0}",
+                readFifo.Count);
+        }
+
+        private void DoProgramData(long busWordAddress){
+            this.Log(
+                LogLevel.Noisy,
+                "OpenTitan_FlashController/DoProgramData: busWordAddress = 0x{0:X}",
+                busWordAddress);
+            uint value = programFifo.Dequeue();
```

Comment:
What if `programFifo` is empty at this point?

I would suggest using `TryDequeue` method and log warning/set error flag/return in case there is no data in the queue.

Reply:
Should it be executed if the operation is not allowed (this matters more than in case of reading)?

Reply:
The recent refactoring efforts have eliminated programFifo